src/rnnlm/rnnlm-decodable-simple-looped.cc

```diff
@@ -0,0 +1,184 @@
+                                       extra_right_context,
+                                       num_sequences,
+                                       &request1, &request2, &request3);
+
+  CompileLooped(rnnlm, opts.optimize_config, request1, request2,
+                request3, &computation);
+  computation.ComputeCudaIndexes();
+  if (GetVerboseLevel() >= 3) {
+    KALDI_VLOG(3) << "Computation is:";
+    computation.Print(std::cerr, rnnlm);
+  }
+}
+
+DecodableRnnlmSimpleLooped::DecodableRnnlmSimpleLooped(
+    const DecodableRnnlmSimpleLoopedInfo &info) :
+    info_(info),
+    computer_(info_.opts.compute_config, info_.computation,
+              info_.rnnlm, NULL),  // NULL is 'nnet_to_update'
+    // since everytime we provide one chunk to the decodable object, the size of
+    // feats_ == frames_per_chunk
+    feats_(info_.frames_per_chunk,
+           info_.word_embedding_mat.NumRows()),
+    current_log_post_offset_(-1)
+{
+  num_frames_ = feats_.NumRows();
+}
+
+void DecodableRnnlmSimpleLooped::TakeFeatures(
+    const std::vector<int32> &word_indexes) {
+  KALDI_ASSERT(word_indexes.size() == num_frames_);
+  std::vector<std::vector<std::pair<MatrixIndexT, BaseFloat> > >
+      pairs(word_indexes.size());
+  for (int32 i = 0; i < word_indexes.size(); i++) {
+    std::pair<MatrixIndexT, BaseFloat> one_hot_index(word_indexes[i], 1.0);
+    std::vector<std::pair<MatrixIndexT, BaseFloat> > row(1, one_hot_index);
+    pairs[i] = row;
+  }
+  SparseMatrix<BaseFloat> feats_temp(feats_.NumCols(), pairs);
+  feats_.Swap(&feats_temp);
+  // resets offset so that AdvanceChunk() would be called in GetOutput() and
+  // GetNnetOutputForFrame() after taking new features
+  current_log_post_offset_ = -1;
+}
+
+void DecodableRnnlmSimpleLooped::GetNnetOutputForFrame(
+    int32 frame, VectorBase<BaseFloat> *output) {
+  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+    AdvanceChunk();
+  output->CopyFromVec(current_nnet_output_.Row(frame -
+                                               current_log_post_offset_));
+}
+
+BaseFloat DecodableRnnlmSimpleLooped::GetOutput(int32 frame, int32 word_index) {
+  KALDI_ASSERT(frame >= 0 && frame < feats_.NumRows());
+  if (frame >= current_log_post_offset_ + current_nnet_output_.NumRows())
+    AdvanceChunk();
+
+//  int32 embedding_dim = info_.word_embedding_mat.NumCols();
+//  int32 num_words = info_.word_embedding_mat.NumRows();
+
+  const CuMatrix<BaseFloat> &word_embedding_mat = info_.word_embedding_mat;
+
+  CuMatrix<BaseFloat> current_nnet_output_gpu;
+  current_nnet_output_gpu.Swap(&current_nnet_output_);
+  const CuSubVector<BaseFloat> hidden(current_nnet_output_gpu,
+                                      frame - current_log_post_offset_);
+  BaseFloat log_prob =
+    VecVec(hidden, word_embedding_mat.Row(word_index));
+//      output_layer->ComputeLogprobOfWordGivenHistory(hidden, word_index);
```

Comment:
it would probably make sense to have an option here to do the correct normalization, just for baseline-ing purposes if nothing else.